### ISSUE TITLE
[FR] allow users to turn off checks

### ISSUE THREAD

**Issue by SteveBronder** [stan-dev/math#1449]
## Description

A couple times we've talked about how expensive all the runtime checks are. In the past we mentioned using some eigen internals to make sure checks only happen once.

Instead, what if we just had wrapped the check methods in `STAN_NO_CHECKS` macro that the user could supply that makes all the checks no-ops. I think I could write a little python script to make the example below

## Example

Consider the following

```cpp
template <typename T_y, typename T_high>
inline void check_less(const char* function, const char* name, const T_y& y,
                       const T_high& high) {
  internal::less<T_y, T_high, is_vector_like<T_y>::value>::check(function, name,
                                                                 y, high);
}
```

We just do

```cpp
#ifndef STAN_NO_CHECKS
template <typename T_y, typename T_high>
inline void check_less(const char* function, const char* name, const T_y& y,
                       const T_high& high) {
  internal::less<T_y, T_high, is_vector_like<T_y>::value>::check(function, name,
                                                                 y, high);
}
#else
template <typename T_y, typename T_high>
inline void check_less(const char* function, const char* name, const T_y& y,
                       const T_high& high) {} 
#endif
```

It does add some macro stuff which is kind of lame, but the pattern itself is pretty simple and idt that horrific.

## Expected Output

A way to not have to do all the checks in a stan program. So then users could run their model a few times with the checks, then when they are pretty sure things are safe they can define the macro and get a faster program

#### Current Version:
v3.0.0

**Comment by bob-carpenter** [comment]
If we do go down this route, the ifdef should be inside the function body to avoid signature duplication.

**Comment by SteveBronder** [comment]
^agree

imo I'd rather not go this route, macros lead to madness. But if we can make it simple and clear what's happening then it's not the worst answer 

@seantalts did you have a way to do this in the new compiler somehow?

**Comment by seantalts** [comment]
Not for most of the checks, not until we rewrite the Math library in Stan language itself (or pull it in via a Lightweight Modular Staging approach, or something similar). 

It would be nice if we could compile a single translation unit with one version of the log_prob that contained the checks and another version that did not. This particular approach wouldn't allow for that, I think... Maybe that's okay. Can you think of anything that would allow that feature? One idea would be something like a default template parameter:

```cpp
template <typename T_y, typename T_high, no_checks=STAN_NO_CHECKS>
inline void check_less(const char* function, const char* name, const T_y& y,
                       const T_high& high) {
  if (!no_checks) {
    internal::less<T_y, T_high, is_vector_like<T_y>::value>::check(function, name,
                                                                 y, high);
  }
}
```

But to use this effectively this seems like it might also need to get threaded through calling functions like `normal_lpdf`.

**Comment by SteveBronder** [comment]
Yeah I thought about the extra template parameter, but then at every call site we need to have at least a `<>` because of some weird C++14 template rules 

```cpp
check_less<>(blah);
```

Since C++14 doesn't allow for empty template values in this context (c++17 does tho')

**Comment by syclik** [comment]
Although this leads to a doubling of function signatures (for the
functions, not the checks), what if we separated out the computation of the
function from the checks themselves. This would shift the function
precondition to expect the input to be valid.

Then a separate function would be to check the arguments. That function
could be turned on or off. Something like: normal_lpdf and
validate_normal_lpdf.

I had been thinking about what dependency injection does (the design
pattern) as a way to handle validation. But C++ isn't great for something
like that.

**Comment by seantalts** [comment]
Yeah, to build on what @syclik said I think a much cleaner option would follow something like the DI pattern - declaratively prescribe what checks should be performed on which inputs (and outputs?) of each function. Then stanc can read in that data and 1) turn them on and off with a single flag and 2) intelligently optimize them even in cases where users are requesting checked code. An example would be using several functions that all check whether a matrix `X` is positive definite - the compiler can emit code to perform that check just once. This is magnified when you have nested math functions.

**Comment by syclik** [comment]
+1!

Some of that will have to be run-time constructed at the time of the
function call, but a lot of it is going to be redundant checking. For
example, any of our transforms that end up with positive values will never
fail a check for positivity.

**Comment by seantalts** [comment]
Ah, didn't realize some would be runtime. Maybe the first pass just ignores those and leaves them in as-is. 

All that said, I wouldn't personally stand in the way of some variant of what @SteveBronder is proposing here - seems like it could be a potentially quicker way to a win, even if it's missing some features from something like this fancier solution we're talking about.

**Comment by SteveBronder** [comment]
Sean or Dan can one of yinz provide a link that explains that approach? It sounds sort of like contracts that we want to be able to enable or disable. Maybe like some class that takes in generic lambas then executes them if the checks are turned on

```cpp
auto check_less = [](auto& x) { check_less_impl(x);};

// pseudoclass for applying checks
template <typename... Checks>
class the_checker {
  // probably wrongish but just a container holding lambdas
  std::tuple<std::function<Checks>...> the_checks;
  // the requires thing is just some yada yada
  template <typename... Args, require_all_not_same_t<the_checker<Args...,> Args...>...>
  the_checker(Args&&... args) {
      // apply the args to each lambda check
      apply_lambas(the_checks, std::forward<Args>(args)...);
  }
}

// Some function that needs checks
void my_danger_func(const Eigen::MatrixXd& x) {
  the_checker<check_less, check_positive, check_danger> x_check(x);
}
```

Though the only positive I see from the above is that any global stuff we apply to turn things on and off will still be global, though isolated which is nice.

**Comment by alashworth** [comment]
Are you guys thinking of some approach where the compiler will hoist/lift function invariants (like positive-definiteness) out of each function block so it's only done once? That makes a lot of sense, especially if you have the full computation graph.

IMO from an outsider perspective I'd vote for a macro solution. It's what Eigen and other similar libraries do. It is harder to understand the template version, and either way you'll need recompilation since my understanding is you don't want duplicate function hanging around (one with checks, one without).

**Comment by syclik** [comment]
@SteveBronder, I was just re-reading this issue. Can we change the focus of the issue to describing the problem rather than jumping to the problem? As it stands, I'd lean towards closing this because I don't think the macro solution actually accomplishes what you're trying to do. If you described the issue, that could stand until we came up with a solution to close it.

**Comment by SteveBronder** [comment]
> Can we change the focus of the issue to describing the problem rather than jumping to the problem?

Sorry I'm not sure if I'm understanding the difference between the two.

Should the issue be changed to, "How do we let users turn on and off checks?" or something of that sort? The initial issue was the proposal for a macro, but this kind of turned into generally what are the ways to handle runtime checks

**Comment by bob-carpenter** [comment]
jumping to the problem -> jumping to the solution.

I see two primary uses of issues.

One is to make a feature request.  I think Daniel's suggesting you turn this into a feature request of the form "allow users to turn off checks".  A feature request should be in functional spec form---that is, it should say what the behavior is from the user side that would allow the issue to be closed.  I think we should allow such feature requests to have technical specs attached if we believe there's only one sensible way to do something. Then completing the issue means doing it according to the technical spec, in whatever detail that is. 

The second major use is a bug report.  For that, the main requirement is reproducibility so we can create a test based on the issue.  Closing the issue happens when the bug's fixed.  Again, I think it's fair to include technical specs on how a bug needs to be fixed if that's important.

So it's really about being definite enough that someone can actually write code for the issue and close it without too much coordination.  I'd rather not have vague issues about general direction that still require design docs before being completed.  

I'd like the issues to reflect tasks that someone can take up and work on without a lot of coordination.  This is because I want to lean heavily toward accepting any pull request that satisfies a feature request.  More functional issues are more open to interpretation.  What I want to avoid is a developer working on an issue only to be told it's not being done the right way.  All within the bounds of our coding standards, of course!

**Comment by syclik** [comment]
Thanks, Bob! That’s exactly how I was thinking about it.

Steve, sorry it wasn’t clear the first time around.

**Comment by bob-carpenter** [comment]
We've drifted to this perspective over time.  It used to be "put it in an issue so we don't forget it".  As the number of issues grew, that became unworkable as triage became really painstaking.

**Comment by SteveBronder** [comment]
I was browsing how gcc works with `-fno-exception`

https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_exceptions.html

I think a clean pretty easy way to do this would be to have a macro `check__` defined as 

```cpp
#ifdef STAN_NOEXCEPT
#define check__ if(false)
#else 
#define check__ if(true)
#endif
```

Then all the checks can just have

```
// in check function body
check__ {
// do check stuff...
}
```

Though it uses a macro that feels nice and mininal

**Comment by bob-carpenter** [comment]
We don't want anying gcc-specific in the code---it should work with all compilers.  gcc's so slow at compilation that I find it unusable (e.g., std::complex tests compile in 45s on clang++ and 16m on g++).

Also, two trailing underscores are technically illegal in user code---it's reserved for system calls.  Though we've used it elsewhere, we shouldn't introduce new uses.

Will this get used in model generated code or just in our internal functions?  If it gets code generated, then we have the issue of potential symbol clashes with generated code and need to make whatever the flag is a reserved word in the Stan language.

**Comment by syclik** [comment]
@SteveBronder, I don't know if you saw this post: https://discourse.mc-stan.org/t/preconditions-and-postconditions-of-functions-in-the-math-library/12127.

I'm starting to think that it may not be so wise to do this. Right now, all our functions, except things in the standard template library, throw an exception or if it doesn't, the output is something that's valid.

If we remove the checks, we may end up in a situation where the output of one function is undefined (because it would have thrown otherwise), but returns a number that feeds into a different function and the error is swallowed. This will cause no end to headaches.

One way to address this properly is to change the Math library completely. If we split each function into a validation piece and a computational piece that had the precondition that the input must be in the domian of the function, then we can just omit the validation piece. But as it stands, I think too many things will actually not work as intended.

One quick way to check whether this is true is to branch Math and make all the `check_*()` no-ops, then go into Stan and check that the `lp` value is not `NaN` before doing the gradient evaluation. If that results in exactly the same samples (down to the last bit) on a reasonably complex model, then I'd believe that this could be done under what you're suggesting for this feature request. If it fails to match, then I think we should think about redesigning the Math library rather than going this route.

**Comment by SteveBronder** [comment]
> We don't want anying gcc-specific in the code---it should work with all compilers.  gcc's so slow at compilation that I find it unusable (e.g., std::complex tests compile in 45s on clang++ and 16m on g++).

Sorry I wasn't clear, the above is just inspired by how gcc does `-fno-exceptions`, I'm not saying we do anything gcc specific

> Also, two trailing underscores are technically illegal in user code---it's reserved for system calls.  Though we've used it elsewhere, we shouldn't introduce new uses.

Ack, I always forget the underscore rules, found them at this [stackoverflow](https://stackoverflow.com/a/3651336/2269255) link

>> 1. identifiers starting with an underscore, followed by a capital character
>> 2. identifiers having two consecutive underscores anywhere
>> 3. identifiers in the global namespace starting with an underscore

> Will this get used in model generated code or just in our internal functions?  If it gets code generated, then we have the issue of potential symbol clashes with generated code and need to make whatever the flag is a reserved word in the Stan language.

This would just be in the internal code. If we wanted to be v v v careful about it we could even have a weird scheme like

> macro_check_def.hpp

```cpp
#ifdef STAN_NOEXCEPT
#define check__ if(false)
#else 
#define check__ if(true)
#endif
```

> macro_check_undef.hpp
```
#undef check__
```

> a_check_thing.hpp

```cpp

#include <stan/...../macro_check_def.hpp>

void some_check(...) {
 check__ {
  // stuff
  }
}

#include <stan/..../macro_check_undef.hpp>
```

Though that requires being super defensive. Maybe the macro solution is too unappetizing 

> @SteveBronder, I don't know if you saw this post: https://discourse.mc-stan.org/t/preconditions-and-postconditions-of-functions-in-the-math-library/12127.

I saw a bit of it, but holidays and work etc. I mostly just creeped on the convo

> I'm starting to think that it may not be so wise to do this. Right now, all our functions, except things in the standard template library, throw an exception or if it doesn't, the output is something that's valid.

> If we remove the checks, we may end up in a situation where the output of one function is undefined (because it would have thrown otherwise), but returns a number that feeds into a different function and the error is swallowed. This will cause no end to headaches.

I 100% agree we should be defensive on behalf of our users, I think that's why a lot of users like Stan, we are pretty good at telling people when they goof. But for our power users whose models can take a _long long_ time I think it would be nice to have a route where, if someone pinky promises they know what they are doing, we are cool with taking the bumpers off the bowling alley.

> One way to address this properly is to change the Math library completely. If we split each function into a validation piece and a computational piece that had the precondition that the input must be in the domian of the function, then we can just omit the validation piece. But as it stands, I think too many things will actually not work as intended.

I'm not sure if I'm following this piece. It sounds like a mathematically rigid style of contracts?

> One quick way to check whether this is true is to branch Math and make all the check_*() no-ops, then go into Stan and check that the lp value is not NaN before doing the gradient evaluation. If that results in exactly the same samples (down to the last bit) on a reasonably complex model, then I'd believe that this could be done under what you're suggesting for this feature request. If it fails to match, then I think we should think about redesigning the Math library rather than going this route.

Maybe we should halt this convo till the next math meeting (or meet up for drinks soon ;-) ). It's certainly a delicate topic, I could see this solution easily failing for something like complicated GPs. But for some simplier models it could be good to look at what the speedup is. If a model takes 2-3 weeks to run, saving someone a day of computing time is pretty nice

**Comment by bob-carpenter** [comment]
If checks are turned off, the only guarantee would be that if the original function succeeded, the unchecked one would too.  The converse would not hold.  The unchecked function may do something different if the original function would throw.

**Comment by SteveBronder** [comment]
> If checks are turned off, the only guarantee would be that if the original function succeeded, the unchecked one would too.  The converse would not hold.  The unchecked function may do something different if the original function would throw.

Yes. To be clear, I'm 100% proposing we hand our users a footgun with a safety being, "knowing how to set macros" ala `-DSTAN_NOCHECKS`

imo it feels like a conversation around contracts, right now we make all attempts to tell our users when they goofed (which is good). I'd like a way for a user to say, "I am willing to sacrifice safety for performance"

**Comment by betanalpha** [comment]
This is the approach that I used in nomad which allows for some drastic 
performance improvements, at least in that optimized autodiff approach.

I think that this is a reasonable UX — the default is the safest approach 
with all of the checks and users wanting no checks have to explicitly set 
the macro, making them responsible for the anything else going forwards.  

To be even safer the interfaces could tag the fit objects/output with a flag 
denoting that no checks were run so that shared output would also 
communicate how it was run.

**Comment by SteveBronder** [comment]
> To be even safer the interfaces could tag the fit objects/output with a flag
denoting that no checks were run so that shared output would also
communicate how it was run.

I like this idea a lot, it also makes it clear to the user that if they see something weird when running the model again having that flag on is a candidate as to why things are goofing

Updating below with proposed implementation

```cpp
// define this macro somewhere
#ifdef STAN_NOEXCEPT
#define check__ if(false)
#else 
#define check__ if(true)
#endif


// sometime l8r...

template <typename Derived>
inline void check_pos_definite(const char* function, const char* name,
                               const Eigen::LDLT<Derived>& cholesky) {
 check__ {
    if (cholesky.info() != Eigen::Success || !cholesky.isPositive()
        || !(cholesky.vectorD().array() > 0.0).all()) {
      domain_error(function, "LDLT decomposition of", " failed", name);
    }
  }
}
```
If people are cool'ish about the above I can do that for the error functions that are used in the functions for the gold checks so we can get a baseline as to how helpful this is

**Comment by syclik** [comment]
Honestly, we can have both: a consistent contract and performance improvements that's unsafe. The way to do this is to refactor the code base. 

1. The existing functions / API would have minimal preconditions. If it doesn't error, then the result is to be trusted. These functions would be split into two components that would be exposed:
2. Functions that compute the value with preconditions that need to be satisfied in order for the output to be trusted. If the preconditions are not satisfied, the result is not to be trusted and it's on the client to check.
3. Functions that validate input. These will throw. And to give the flexibility you're looking for, we can have this function be templated or a macro to drop this out if necessary.

For example, instead of a single `normal_lpdf()` function, we could have 3 functions:
1. `normal_lpdf()` <- same API as today
2. `normal_lpdf_compute()` <- has preconditions
3. `normal_lpdf_check()` <- throws if the preconditions for `normal_lpdf_compute()` are not met.

Then we'd have the flexibility to check or not check when we wanted. Maybe after a while, we actually code generate with the compute and check functions instead of the one that does it all, but who knows.

What do you think?

**Comment by syclik** [comment]
@SteveBronder, what are "gold checks"?

**Comment by SteveBronder** [comment]
> Then we'd have the flexibility to check or not check when we wanted. Maybe after a while, we actually code generate with the compute and check functions instead of the one that does it all, but who knows.

Is the goal with this implementation that users could turn off and on checks for a precompiled model? Would all 3 be exposed to users? I'm not totally understanding how end users would use the above

related, @seantalts would it be possible for us to have a map for each math function s.t. the generate code could be something like

```cpp
if (do_checks) {
  check_something(...);
  check_something_else(...);
}
function_that_uses_checks(...);
```

Then the checks are extracted out of the math library and handled by the compiler. Dan I'm not sure but I feel like the above sort of flows with your idea. We would need to add another parameter flag to the model `do_checks` or whatever we want to call it. And compile a dictionary of checks for each function which would be kind of annoying. If the compiler can look at the code in the math library maybe we could write a script that updates those automatically by looking at comments below the function signature like

```cpp
auto some_func(MatrixXd A, int n) {
  // check_size(A.size(), n);
  // check_pos_def(A);
// ...rest of func
}
```

Or idk, then you also have to do some weird mapping from argument signatures to their real object name in the model which feels complicated and lame

> @SteveBronder, what are "gold checks"?

Meant the stat_comp benchmarks

**Comment by bob-carpenter** [comment]
> 	• normal_lpdf() <- same API as today
> 	• normal_lpdf_compute() <- has preconditions
> 	• normal_lpdf_check() <- throws if the preconditions for normal_lpdf_compute() are not met.

Would users write foo_lpdf_compute(y | theta) in their Stan code?  Is there a way to handle sampling statements like y ~ foo(theta)?

**Comment by bob-carpenter** [comment]
It's not just distributions, it's literally every function in Stan that can potentially do checks.  That's a lot of code generation and hudreds of new functions to code, test, and document.

**Comment by SteveBronder** [comment]
> It's not just distributions, it's literally every function in Stan that can potentially do checks.  That's a lot of code generation and hudreds of new functions to code, test, and document.

Yes now reading both I don't like either mine or Dan's new proposals. imo I like my dumb macro approach

**Comment by syclik** [comment]
(I'm not replying in the order received cause I think some of the comments overlap in theme, but not in time.)

> Yes now reading both I don't like either mine or Dan's new proposals. imo I like my dumb macro approach

I would like to preregister my objection to the macro approach. My honest belief is that if we tried this with any Stan model that's non-trivial will end up with different results. By different results, I mean that if you ran with the same seed, not all the output would be identical. What this means is within that run, there exists a set of input where it the function would have thrown but with the checks off, it didn't. The output of that function was then used and now the run has done something different. To me, this means that the run is actually invalid... we weren't faithfully able to compute the function value, but we continued to use it to generate more MCMC draws. It's not the difference of starting it with another seed. This is the difference of using a function that doesn't return the correct function value. 

This can be shown technically whether this is true or not. So, before including a macro that can allow this behavior, I'd like to see if my objection is true. If it is, the macro seems like a bad option for us, as developers, to expose to clients of this library. If this isn't true and a lot of these models do complete with identical output, then I think the macro option is a decent one.

I understand that adding a macro approach allows us to scale to the whole library faster, but I think we expose ourselves to a whole lot of pain in explaining to everyone that they shouldn't do this unless under extreme circumstances. And honestly, if they're that willing to do it, they could do it now by altering the source.

> related, @seantalts would it be possible for us to have a map for each math function s.t. the generate code could be something like
> 
> ```c++
> if (do_checks) {
>   check_something(...);
>   check_something_else(...);
> }
> function_that_uses_checks(...);
> ```

I was thinking of something far simpler and easier on the Stan compiler. As you've written that, you've left a leaky abstraction... stanc needs to know the internals of the Math function in order to have this map. Rather than that, it could be encapsulated by the Math library. So... going from this (this isn't exactly how it translates in C++, but the idea is right):
```c++
lp__ += normal_lpdf(y, mu, sigma);
```

to 
```c++
if (do_checks)
  normal_lpdf_check(y, mu, sigma);
lp__ += normal_lpdf_compute(y, mu, sigma);
```

Now the burden is on the Math library to update that check function appropriately and stanc doesn't need to know exactly how to check its inputs, `y`, `mu`, and `sigma`. (At some point, it can to do more compiler optimizations, but it doesn't need to.)

> It's not just distributions, it's literally every function in Stan that can potentially do checks. That's a lot of code generation and hudreds of new functions to code, test, and document.

Yes, and I'm aware of that when I'm suggesting this. I don't think the testing and coding will be that difficult. It'll be some work, though. Also, the nice thing is that we don't need to have all the functions split out this way at once. It can be built up slowly without the API being left in a bad place. It will still be functional if we don't do this.

> Would users write foo_lpdf_compute(y | theta) in their Stan code? Is there a way to handle sampling statements like y ~ foo(theta)?

I wasn't actually thinking that far ahead, but YES! This would be on the stanc compiler to know what's available, though. At least until the whole library can do this properly. That way the user would have the choice to call the compute function directly if they choose to.

---

One additional point to make... it would be better to have this sort of separation anyway. Right now, if a function delegates its computation to another, we have no way of turning checks off. Even with the macros. But we might be able to guarantee from within the outer function that we've met preconditions of the second function and we can go directly to the computation. I'm sure we do this in the Math library, but I can't find an example off the top of my head (I was thinking `neg_binomial_2`, but it looks like a complete reimplementation).

**Comment by SteveBronder** [comment]
> This can be shown technically whether this is true or not. So, before including a macro that can allow this behavior, I'd like to see if my objection is true. If it is, the macro seems like a bad option for us, as developers, to expose to clients of this library. If this isn't true and a lot of these models do complete with identical output, then I think the macro option is a decent one.

In my last comment I used the word "cool'ish" because I'm also uncertain if it would work properly. The [sir](https://github.com/stan-dev/stat_comp_benchmarks/tree/64b6597cd17de77e91dcfb1a9977758407017d20/benchmarks/sir) model is very angsty and would be a good sanity check candidate

>```
>lp__ += normal_lpdf(y, mu, sigma);
>```
>to
>```
>if (do_checks)
>  normal_lpdf_check(y, mu, sigma);
> lp__ += normal_lpdf_compute(y, mu, sigma);
>```
> Now the burden is on the Math library to update that check function appropriately and stanc doesn't need to know exactly how to check its inputs, y, mu, and sigma. (At some point, it can to do more compiler optimizations, but it doesn't need to.)

So, I think what I'm missing here is how, to the user, is this different from the macro version? Here users either have set `do_check` to `true` or `false` right? Or would `do_checks` be turned on/off in some clever way? Cuz if it's set at runtime then it feels like all the stuff you have above is still true for this approach as well.

**Comment by bob-carpenter** [comment]
@syclik --- you should just close the issue and mark it will-not-fix if you are not willing to put up with differing behavior when checks are removed.  As @syclik points out, using a conditional has exactly the same problem of differing behavior with checks on and off for some program/data combos.

> stanc needs to know the internals of the Math function

I was imagining @syclik's version where the tests are named after the function as we do for `_lpdf` functions and sampling statements already.  But that doesn't solve the differing behavior, it'll just be way more complicated to implement and plumb through.

> The sir model is very angsty

What does "angsty" mean in this context, @SteveBronder?

> if they're that willing to do it, they could do it now by altering the source.

This is just as true (or false) of any feature we might add to Stan.  Very few of our users have the expertise to make major changes to C++ code.  This particular change doesn't require deep knowledge of C++, but it is a huge amount of work to do thoroughly.

**Comment by syclik** [comment]
> @syclik --- you should just close the issue and mark it will-not-fix if you are not willing to put up with differing behavior when checks are removed.

Sorry if this has gotten you frustrated. That isn't what I've said and I'll try to clarify.

- I won't close this issue because as a feature request it should stand. We want faster functions without checks. That makes sense to me. (We're talking about how to do this, not whether we should.)
- If the issue was tied to the proposed implementation of macros, then I don't have enough information to close it now. I'd have enough information if we could tell if this is worth considering. It's only after we evaluate the models that we can really determine if we should or should not do this. Right now, it's a good thought. If there's a prototype that actually works, then we can determine whether it's worth even considering putting up with differing behavior.

> So, I think what I'm missing here is how, to the user, is this different from the macro version?

@SteveBronder, there are a few things:
- using a macro: a function that throws will behave differently if a C preprocessor macro is defined. 
- using a different function: the function is unaffected C prepreprocessor macros. It's on the user to know which function they want to call; either the one with checks or one without.

It would really help a different user, though. Math developers could now use the check-less functions for implementation instead of only using the functions with checks. Clients of the Math library would have to figure it out somehow.

> This is just as true (or false) of any feature we might add to Stan. Very few of our users have the expertise to make major changes to C++ code. This particular change doesn't require deep knowledge of C++, but it is a huge amount of work to do thoroughly.

Ok. This one isn't a huge amount of work... there are 57 `check_*.hpp` files. In each of those, if the first line in the function is `return;` it'll knock out the check. The knowledge is harder than the work for this particular example. Of course, if someone knew their exact model, they could reduce the number of files to touch by a lot.

But I think the question we should answer (as a group) is whether we want to expose this behavior to users and under what conditions. I think there's a use for this as a developer developing a function calling other functions regardless of how we expose this to clients of the library. But under that condition, we can't really just turn off all checks to make that work.

**Comment by syclik** [comment]
For testing purposes, I put up a branch that eliminates all the code in the check functions: `feature/noop-checks`.

@SteveBronder, want to check this on that model? See if we can get the same results?

**Comment by rok-cesnovar** [comment]
Nice!

I started a Cmdstan Performance Test with your branch @syclik : https://jenkins.mc-stan.org/blue/organizations/jenkins/CmdStan%20Performance%20Tests/detail/Custom/138/pipeline

Will update with the results once that finishes.

**Comment by SteveBronder** [comment]
Thanks guys! Dan looking at what Rok is running I think you may have missed a few? Some of the benchmarks are still throwing

**Comment by syclik** [comment]
> Thanks guys! Dan looking at what Rok is running I think you may have missed a few? Some of the benchmarks are still throwing

@rok-cesnovar, I'm not sure that's the right branch. One of the checks that pop up:
```
Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:

Exception: gp_exp_quad_cov: length_scale is 0, but must be > 0! (in '../stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan', line 16, column 4 to line 17, column 59)
```

When I look at `stan/math/opencl/prim/gp_exp_quad_cov.hpp`, it calls `check_positive_finite()` with `length_scale`. Tracing through `check_positive_finite()`, it should return and not do what it's doing. I've verified by running these unit tests to see that they all fail:
- test/unit/math/rev/mat/err/check_positive_finite_test.cpp
- test/unit/math/rev/arr/err/check_positive_finite_test.cpp
- test/unit/math/rev/scal/err/check_positive_finite_test.cpp
- test/unit/math/prim/mat/err/check_positive_finite_test.cpp
- test/unit/math/prim/arr/err/check_positive_finite_test.cpp
- test/unit/math/prim/scal/err/check_positive_finite_test.cpp

They all compile and run and as expected, the unit tests fail because nothing throws.

**Comment by seantalts** [comment]
> related, @seantalts would it be possible for us to have a map for each math function s.t. the generate code could be something like

Definitely - architecturally I think that's the best answer here, but it requires some refactoring. Even with checks on, the compiler can then optimize and make sure we only check data once, and check parameters only once per iteration.

**Comment by rok-cesnovar** [comment]
@SteveBronder @syclik doesn't this performance test also run the tests on develop cmdstan? Could those throws be coming from that? Anyway, I re-ran the perf tests with some more iterations.

The results from that run were this on Linux: 

| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) | 
| ------------- |------------- | ------------- | ------------- | ------------- | 
| stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan | 4.72 | 4.29 | 1.1 | 9.18% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan | 0.01 | 0.01 | 1.01 | 0.53% faster | 
| stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan | 0.07 | 0.07 | 1.01 | 1.03% faster | 
| stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan | 9.86 | 10.14 | 0.97 | -2.88% slower | 
| compilation | 51.64 | 50.6 | 1.02 | 2.0% faster | 
| stat_comp_benchmarks/benchmarks/arK/arK.stan | 2.07 | 2.06 | 1.01 | 0.6% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan | 10.31 | 9.47 | 1.09 | 8.18% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan | 13.69 | 13.68 | 1.0 | 0.1% faster | 
| stat_comp_benchmarks/benchmarks/sir/sir.stan | 80.16 | 80.13 | 1.0 | 0.05% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan | 0.27 | 0.28 | 1.0 | -0.13% slower | 
| stat_comp_benchmarks/benchmarks/garch/garch.stan | 0.61 | 0.61 | 1.0 | 0.33% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan | 0.18 | 0.18 | 1.05 | 4.48% faster | 
| stat_comp_benchmarks/benchmarks/arma/arma.stan | 0.79 | 0.78 | 1.02 | 1.66% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan | 0.03 | 0.03 | 1.03 | 2.69% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan | 3.82 | 3.48 | 1.1 | 8.88% faster | 
Mean result: 1.02643872587

and for Mac:

| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) | 
| ------------- |------------- | ------------- | ------------- | ------------- | 
| stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan | 4.99 | 4.61 | 1.08 | 7.53% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan | 0.02 | 0.02 | 1.0 | -0.02% slower | 
| stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan | 0.1 | 0.09 | 1.08 | 7.16% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan | 0.23 | 0.21 | 1.07 | 6.42% faster | 
| compilation | 88.32 | 84.93 | 1.04 | 3.84% faster | 
| stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan | 6.21 | 5.82 | 1.07 | 6.3% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan | 7.49 | 6.4 | 1.17 | 14.61% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan | 20.23 | 21.13 | 0.96 | -4.45% slower | 
| stat_comp_benchmarks/benchmarks/sir/sir.stan | 92.9 | 101.66 | 0.91 | -9.44% slower | 
| stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan | 0.05 | 0.04 | 1.03 | 2.5% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan | 2.98 | 2.58 | 1.16 | 13.49% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan | 0.32 | 0.33 | 0.95 | -5.46% slower | 
| stat_comp_benchmarks/benchmarks/arK/arK.stan | 2.35 | 1.76 | 1.33 | 25.0% faster | 
| stat_comp_benchmarks/benchmarks/arma/arma.stan | 0.8 | 0.77 | 1.03 | 2.86% faster | 
| stat_comp_benchmarks/benchmarks/garch/garch.stan | 0.53 | 0.62 | 0.86 | -16.74% slower | 
Mean result: 1.04835870385

The Windows instance doesnt currently output this nice table (Nic told me its some weird Jenkins Windows bug he is looking into) and I didnt bother parsing it manually because the result for Linux and Mac arent all that convincing. But as I said there werent many iterations there. Running with 10 iterations on the gold tests now: https://jenkins.mc-stan.org/blue/organizations/jenkins/CmdStan%20Performance%20Tests/detail/Custom/140/pipeline

**Comment by SteveBronder** [comment]
I'll look at that branch tonight. I think some of the checks are still in there because the checks for the branch are still throwing stuff like below

```
Exception: cholesky_decompose: Matrix m is not positive definite (in '../stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan', line 16, column 2 to column 47)
```

**Comment by SteveBronder** [comment]
I jus threw up a branch that I'm pretty sure fixes those, running tests here but this one still has the ode integrator checks for max iterations. Have a local version running that removes those. I would guess removing those may be v v math wrong

https://jenkins.mc-stan.org/job/CmdStan%20Performance%20Tests/job/Custom/141/console

**Comment by SteveBronder** [comment]
>> The sir model is very angsty
> What does "angsty" mean in this context, @SteveBronder?

Just that it tends to throw on the first few iterations and is one of the more complex models in the suite

**Comment by bob-carpenter** [comment]
> Sorry if this has gotten you frustrated. 

Not at all.  I don't care much about this feature one way or the other.  It doesn't seem worth the effort to me, but I'm not going to try to block it if others think it's worthwhile.  

I just think that if differing behavior is a deal-breaker, this issue isn't going to fly as written.  The issue body *is* about using a macro to turn off checks:

> wrapped the check methods in `STAN_NO_CHECKS` macro that the user could supply that makes all the checks no-ops.

Adding two new functions (checks and body w/o checks) for each existing function feels like a very different proposal.  This issue has devolved into a wide-ranging design discussion, which is why I'm suggesting closing it and moving the discussion elsewhere.  

Splitting functions is going to be a pain if the use case is to have modelers start with the checked functions then remove the checks.  A macro would make that easy.

> I think there's a use for this as a developer developing a function calling other functions regardless of how we expose this to clients of the library.

That would certainly be useful to eliminate checks inside nested calls.  The right approach to coding checks in an API is to check user-facing function inputs, then manage the pre-/post-conditions without checks internally after that.  As you point out, we can't do that now because most of our functions are user facing and have checks built-in.

> The knowledge [of C++] is harder than the work for this particular example.

Right---most of our users don't know C++, how to build Stan, etc.  I'd hazard most have never even worked with a compiled language, with make, and many have probably never worked with the terminal.

**Comment by bob-carpenter** [comment]
My guess is that pretty much every situation that would throw would lead to an infinite or NaN result that would get rejected anyway.

> Just that it tends to throw on the first few iterations and is one of the more complex models in the suite

**Comment by rok-cesnovar** [comment]
I agree that refactoring the entire Math library to have optional checks is probably not worth the investment. Especially as it would be kind of a "experimental" mode kind of thing. But if someone is willing to do it, why not.

What would be more worth exploring is how to avoid repeated checks on data in a Stan model. We could get some speedups there without going in the "danger" zone where we have to sacrifice something for speed.

**Comment by SteveBronder** [comment]
Tests just finished 

## Linux:

| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) | 
| ------------- |------------- | ------------- | ------------- | ------------- | 
| stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan | 0.01 | 0.01 | 0.97 | -3.54% slower | 
| stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan | 4.68 | 4.32 | 1.08 | 7.7% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan | 0.28 | 0.28 | 1.0 | 0.06% faster | 
| stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan | 0.07 | 0.06 | 1.02 | 1.55% faster | 
| stat_comp_benchmarks/benchmarks/arK/arK.stan | 2.07 | 2.08 | 1.0 | -0.32% slower | 
| compilation | 51.73 | 50.1 | 1.03 | 3.15% faster | 
| stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan | 10.36 | 10.22 | 1.01 | 1.41% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan | 10.29 | 9.43 | 1.09 | 8.41% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan | 13.86 | 13.5 | 1.03 | 2.55% faster | 
| stat_comp_benchmarks/benchmarks/sir/sir.stan | 79.78 | 78.19 | 1.02 | 1.99% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan | 0.03 | 0.03 | 1.02 | 1.51% faster | 
| stat_comp_benchmarks/benchmarks/garch/garch.stan | 0.63 | 0.56 | 1.12 | 10.44% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan | 0.18 | 0.17 | 1.06 | 5.62% faster | 
| stat_comp_benchmarks/benchmarks/arma/arma.stan | 0.78 | 0.78 | 1.01 | 0.8% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan | 3.81 | 3.48 | 1.1 | 8.71% faster | 
Mean result: 1.03616351148

## Mac:

| Name | Old Result | New Result | Ratio | Performance change( 1 - new / old ) | 
| ------------- |------------- | ------------- | ------------- | ------------- | 
| stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan | 5.16 | 4.65 | 1.11 | 9.88% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan | 0.02 | 0.02 | 0.98 | -1.94% slower | 
| stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan | 0.1 | 0.1 | 1.0 | -0.08% slower | 
| stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan | 0.22 | 0.21 | 1.07 | 6.77% faster | 
| compilation | 87.02 | 84.39 | 1.03 | 3.02% faster | 
| stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan | 6.24 | 5.79 | 1.08 | 7.1% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan | 7.54 | 6.54 | 1.15 | 13.26% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan | 20.5 | 21.93 | 0.93 | -6.98% slower | 
| stat_comp_benchmarks/benchmarks/sir/sir.stan | 99.61 | 98.65 | 1.01 | 0.96% faster | 
| stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan | 0.05 | 0.04 | 1.02 | 2.08% faster | 
| stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan | 2.97 | 2.59 | 1.15 | 12.83% faster | 
| stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan | 0.31 | 0.31 | 1.01 | 1.18% faster | 
| stat_comp_benchmarks/benchmarks/arK/arK.stan | 2.33 | 1.75 | 1.33 | 25.01% faster | 
| stat_comp_benchmarks/benchmarks/arma/arma.stan | 0.8 | 0.75 | 1.06 | 6.0% faster | 
| stat_comp_benchmarks/benchmarks/garch/garch.stan | 0.53 | 0.62 | 0.85 | -17.54% slower | 
Mean result: 1.05304762187

Seems really weird that garch is 18% slower? Also pretty surprised arK is 25% faster. I think we need to start reporting standard deviations for these as well

If it's really only a 5% bump tho I'm not sure how much effort this is worth. Maybe we should push this back until we have better performance testing

**Comment by seantalts** [comment]
None of these test large data, where the checks are most noticeable.

**Comment by bob-carpenter** [comment]
> But if someone is willing to do it, why not.

The usual reason to not include a feature even if it is free or very low cost to develop is to avoid longer-term maintenance costs and to keep things simple.  More functions complicates doc, as do more options.  The tradeoff needs to be considered for every feature.  

But as I said, I'm not trying to argue either way on this one---it doesn't matter much to me one way or the other so I'll be happy either way it goes.

**Comment by bob-carpenter** [comment]
They may be running more iterations without checks.  Failed checks cause exceptions and instant termination of NUTS tree building.

Benchmarking the log density evaluation is the way to go longer term for this kind of thing as it doesn't have any randomness in it.

**Comment by syclik** [comment]
Did we every verify if the runs without checks resulted in identical output? If it didn't, then we've gotten 5% faster on incorrect runs, which isn't really that useful.

**Comment by SteveBronder** [comment]
It checks at 10e-8 by default

**Comment by syclik** [comment]
Awesome! This goes against my intuition, but I think the edge cases are handled well enough where we could probably forgo it for speed.

Of course, we may want to check for `NaN` values of the `lp` and reject immediately. That might make the performance more drastic.

I think @bob-carpenter is correct... under certain cases, the tree keeps building through the end of the treedepth iteration. I would have thought that the RNG would have drifted, but I haven't seen an actual example of that happening, so maybe we're safe? I'm sure I could hand-craft an example that fails, but that would take effort.

I think that's enough information for me. If you think you can something like this so it's easily maintained, go for it.

**Comment by SteveBronder** [comment]
> I think that's enough information for me. If you think you can something like this so it's easily maintained, go for it.

Word! Are people cool with the macro approach or are there other patterns people want to discuss?

**Comment by rok-cesnovar** [comment]
Done with #2423
